src/pyodide/internal/asgi.py

```diff
@@ -0,0 +1,176 @@
+from asyncio import Future, ensure_future, Queue
+from inspect import isawaitable
+from contextlib import contextmanager
+
+ASGI = {"spec_version": "2.0", "version": "3.0"}
+
+
+@contextmanager
+def acquire_js_buffer(pybuffer):
+    from pyodide.ffi import create_proxy
+
+    px = create_proxy(pybuffer)
+    buf = px.getBuffer()
+    px.destroy()
+    try:
+        yield buf.data
+    finally:
+        buf.release()
+
+
+def request_to_scope(req, ws=False):
+    from js import URL
+
+    headers = [tuple(x) for x in req.headers]
+    url = URL.new(req.url)
+    scheme = url.protocol[:-1]
```

Comment:
I'd add a couple asserts in here

```suggestion
    assert url.protocol[-1] == ':'
    scheme = url.protocol[:-1]
```